src/net_processing.cpp

```diff
@@ -3081,8 +3156,46 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
     }
 
     if (strCommand == NetMsgType::NOTFOUND) {
-        // We do not care about the NOTFOUND message, but logging an Unknown Command
-        // message would be undesirable as we transmit it ourselves.
+        // Remove the NOTFOUND transactions from the peer
+        LOCK(cs_main);
+        CNodeState *state = State(pfrom->GetId());
+        std::vector<CInv> vInv;
+        vRecv >> vInv;
+        if (vInv.size() <= MAX_INV_SZ) {
+            for (CInv &inv : vInv) {
+                if (inv.type == MSG_TX || inv.type == MSG_WITNESS_TX) {
+                    state->m_tx_download.m_tx_announced.erase(inv.hash);
+                    auto in_flight_it = state->m_tx_download.m_tx_in_flight.find(inv.hash);
+                    if (in_flight_it == state->m_tx_download.m_tx_in_flight.end()) {
+                        // Ignore NOTFOUND messages unless we requested the
+                        // transaction from the peer, to prevent misbehaving peers from
+                        // polluting the NOTFOUND queues or triggering multiple
+                        // concurrent downloads of the same transactions
+                        continue;
+                    }
+                    state->m_tx_download.m_tx_in_flight.erase(inv.hash);
+                    EraseTxRequest(inv.hash);
+                    // If we don't already have this transaction, check to see
+                    // if any of our outbound peers have offered it to us, and
+                    // if so signal them to try to request immediately. This avoids
+                    // waiting for the download timeout to expire before
+                    // re-requesting a transaction.
+                    if (!AlreadyHave(inv)) {
+                        for (auto it = g_outbound_peers.begin(); it != g_outbound_peers.end(); ++it) {
+                            if (*it == pfrom->GetId()) continue;
+                            CNodeState *state = State(*it);
```

Comment:
This `state` shadows the `state` in the outer scope.

Reply:
Fixed.